dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp

```diff
@@ -101,80 +107,94 @@ PageDirectoryPtr PageDirectoryFactory::createFromEdit(String storage_name, FileP
 
 void PageDirectoryFactory::loadEdit(const PageDirectoryPtr & dir, const PageEntriesEdit & edit)
 {
-    PageDirectory::MVCCMapType & mvcc_table_directory = dir->mvcc_table_directory;
-
     for (const auto & r : edit.getRecords())
     {
-        if (auto it = max_apply_page_ids.find(r.page_id.high); it == max_apply_page_ids.end())
-        {
-            max_apply_page_ids[r.page_id.high] = r.page_id.low;
-        }
-        else
-        {
-            it->second = std::max(it->second, r.page_id.low);
-        }
-
         if (max_applied_ver < r.version)
             max_applied_ver = r.version;
         max_applied_page_id = std::max(r.page_id, max_applied_page_id);
 
-        auto [iter, created] = mvcc_table_directory.insert(std::make_pair(r.page_id, nullptr));
-        if (created)
+        // We can not avoid page id from being reused under some corner situation. Try to do gcInMemEntries
+        // and apply again to resolve the error.
+        if (bool ok = applyRecord(dir, r, /*throw_on_error*/ false); unlikely(!ok))
         {
-            iter->second = std::make_shared<VersionedPageEntries>();
+            dir->gcInMemEntries();
+            applyRecord(dir, r, /*throw_on_error*/ true);
+            LOG_FMT_INFO(DB::Logger::get("PageDirectoryFactory"), "resolve from error status done, continue to restore");
```

Comment:
`DB::Logger::get("PageDirectoryFactory")` add a log in PageDirectoryFactory to replace it? cause it used many place

Reply:
Now we only create it when *failover* does happen.
Just a small optimization to reduce the size of global `_pLoggerMap` in Poco. If it does not occur, we won't call `Poco::Logger::get` and we won't add a new element to the global `_pLoggerMap`. 

https://github.com/pingcap/poco/blob/e411ea34492bf26e76b4e4a03ff5813a0c779240/Foundation/src/Logger.cpp#L475-L480